File: AOC_22/aoc_07_pt1.js
const fs = require('fs')

const readFileLines = filename =>
fs 
  .readFileSync(filename)
  .toString('UTF8')
  .split('\n')

const lines = readFileLines('input_07.txt');
lines.pop();
let current_directory = "/";
let i = 0;

const directories = {"/":0};
const parent_directories = {"/":[]};
while( i < lines.length) {
  let new_line = lines[i];
  if (new_line[0] == "$"){
    new_line = new_line.split(' ');
    if (new_line[1] == "cd"){
      current_directory = new_line[2];
      i++;
    } else{
      i++;
      new_line = lines[i];
      while (i < lines.length && new_line[0] != "$"){
        new_line = new_line.split(' ');
        if (new_line[0] == "dir"){
          if (parent_directories.hasOwnProperty(current_directory)){
            let old_list = parent_directories[current_directory];
            old_list.push(new_line[1]);
            parent_directories[current_directory] = old_list;
          } else{
            parent_directories[current_directory] = [new_line[1]];
          }
        } else {
          let new_val = parseInt(new_line[0]);
          if (directories.hasOwnProperty(current_directory)){
            let old_val = directories[current_directory];
            old_val+=new_val;
            directories[current_directory] = old_val;
          } else{
            directories[current_directory] = new_val;
          }
        }
        i++;
        new_line = lines[i];
      }
    }
  }
}

function getDirectorySize(dir,base_dir, parents){
  let sz = base_dir[dir];

  if (parents.hasOwnProperty(dir)){
    let child_dirs = parents[dir];
    for (let new_dir of child_dirs){
      sz += getDirectorySize(new_dir,base_dir,parents);
    }
  }
  return sz
}


let ans = 0;
let cur_total = 0;
console.log(parent_directories)
for ( let key in directories){
  let val = (getDirectorySize(key,directories,parent_directories));
  if (val <=100000){
    ans += val;
  }
}
console.log(ans);

File: AOC_22/aoc_07_pt1.py
file = open('input_07.txt','r')
lines = file.readlines()



directories = {}
parents = {}

current_directory = "/"
i = 0
while i < len(lines):
    line = lines[i].strip().split(' ')
    print(line)
    # if command
    if line[0] == "$":
        if line[1] == "cd":
            if line[2] == "..":
                for key, value in parents.items():
                    if value is not None and current_directory in value:
                        current_directory = key
                        break
            else:
                next_directory = line[2]
                if current_directory in parents.keys():
                    old_list = parents.get(current_directory)
                    old_list.append(next_directory)
                    old_list = list(set(old_list))
                    parents.update({current_directory : old_list})
                else:
                    parents.update({current_directory : [next_directory]})

                current_directory = next_directory

            i += 1
        else:
            i+=1
            while i < len(lines) and lines[i][0] != "$":
                line = lines[i].strip().split(' ')
                if line[0] == "dir":
                    if current_directory in parents.keys():
                        old_list = parents.get(current_directory)
                        old_list.append(line[1])
                        old_list = list(set(old_list))
                        parents.update({current_directory : old_list})
                    else:
                        parents.update({current_directory : [line[1]]})
                else:
                    if current_directory in directories.keys():
                        old_val = directories.get(current_directory)
                        old_val += int(line[0])
                        directories.update({current_directory: old_val})
                    else:
                        directories.update({current_directory: int(line[0])})
                i += 1






def dirSize(dir,direct,rents):
    size = 0
    if dir in directories.keys():
        size += directories.get(dir)
    if dir in rents.keys():
        children = rents.get(dir)
        for child in children:
            if child != dir:
                size += dirSize(child,direct,rents)
    return size
print(directories)
print(parents)
ans = 0
for key, value in directories.items():
    sz = dirSize(key,directories,parents)
    if sz <= 100000:
        ans +=sz
print(ans)
